### PR TITLE
CADL manual API review

### PR DESCRIPTION
**Pull request by praveenkuttappan**
Create new pipeline to generate API review from CADL file and APIView changes to support CADL

### CODE REVIEW

File: eng/pipelines/apiview-review-gen-cadl.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      versionSpec: $(NodeVersion)
+    displayName: "Use Node $(NodeVersion)"
+
+  - pwsh: |
+      npm install -g cadl
+      npm install -g @cadl-lang/compiler
+    displayName: "Install cadl compiler"
+
+  - pwsh: |
+      Write-Host "Review Details Json: $(Reviews)"
```
Comment (weshaggard) [inline eng/pipelines/apiview-review-gen-cadl.yml:31]: Where is this Reviews variable coming from? Should it be a runtime parameter to this pipeline? What happens if this is not set or set incorrectly?

File: eng/pipelines/apiview-review-gen-cadl.yml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      SkipCheckoutNone: true
+      Repositories:
+      - Name: $(RepoName)
+        Commitish: $(BranchName)
+        WorkingDirectory: $(System.DefaultWorkingDirectory)/cadlRepo
+      Paths:
+      - $(PackagePath)
+
+  - pwsh: |
+      npm install
```
Comment (weshaggard) [inline eng/pipelines/apiview-review-gen-cadl.yml:53]: Should we combine all the npm install steps?

File: eng/pipelines/apiview-review-gen-cadl.yml (hunk 1 above)
Comment (weshaggard) [reply eng/pipelines/apiview-review-gen-cadl.yml:31]: Also based on the (s) at the end I'd assume you could give multiple reviews but this step seems to only handle a single review.

File: src/dotnet/APIView/APIViewWeb/Client/src/reviews.ts [hunk 3]
```diff
@@ -126,4 +127,26 @@
...
+  languageSelect.on('change', function (e) {
+    var val = $(this).val();
+    switch (val) {
+      case 'C#':
+        val = "Csharp";
+        break;
+      case 'C++':
+        val = "Cpp";
+        break;
+    }
```
Comment (benbp) [inline src/dotnet/APIView/APIViewWeb/Client/src/reviews.ts:141]:
```suggestion
    var val = $(this).val().replace("C++", "Cpp").replace("C#", "Csharp");
```

Would be a little more concise, if you want that.

File: src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs [hunk 4]
```diff
@@ -721,12 +729,12 @@ private async Task UpdateReviewAsync(ReviewModel review)
                 review.ServiceName = p?.ServiceName ?? review.ServiceName;
             }
 
-            var languageService = _languageServices.FirstOrDefault(s => s.IsSupportedFile(name));
-            //Run pipeline to generate the review if sandbox is enabled
+            var languageService = language != null ? _languageServices.FirstOrDefault( l=> l.Name == language) : _languageServices.FirstOrDefault(s => s.IsSupportedFile(name));
+            //Run pipeline to generateteh review if sandbox is enabled
```
Comment (benbp) [inline src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs:733]:
```suggestion
            // Run pipeline to generate the review if sandbox is enabled
```

File: src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs [hunk 5]
```diff
@@ -747,12 +755,13 @@ private async Task UpdateReviewAsync(ReviewModel review)
...
-            bool runAnalysis)
+            bool runAnalysis,
+            string language)
         {
             using var memoryStream = new MemoryStream();
-            var codeFile = await CreateCodeFile(originalName, fileStream, runAnalysis, memoryStream);
+            var codeFile = await CreateCodeFile(originalName, fileStream, runAnalysis, memoryStream, language);
             var reviewCodeFileModel = await CreateReviewCodeFileModel(revisionId, memoryStream, codeFile);
-            reviewCodeFileModel.FileName = originalName;
+            reviewCodeFileModel.FileName = originalName;
```
Comment (benbp) [inline src/dotnet/APIView/APIViewWeb/Managers/ReviewManager.cs:764]:
```suggestion
            reviewCodeFileModel.FileName = originalName;
```

File: eng/pipelines/apiview-review-gen-cadl.yml (hunk 1 above)
Comment (praveenkuttappan) [reply eng/pipelines/apiview-review-gen-cadl.yml:31]: Yes. I would like to discuss this with you. I am processing only 1 review to utilize sparse checkout and solution involves changing sandboxing framework for all languages.

Comment (praveenkuttappan) [reply eng/pipelines/apiview-review-gen-cadl.yml]: Fixed to support multiple reviews

File: eng/pipelines/apiview-review-gen-cadl.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+  - pwsh: |
+      $url = "${{parameters.APIViewURL}}"
+      echo "##vso[task.setvariable variable=APIViewURL]$url"
+    displayName: "Setup APIViewURL Variable"
+    condition: eq(variables['APIViewURL'], '')
+
+  - pwsh: |
+      npm install -g @azure-tools/cadl-apiview
+      npm install -g cadl
```
Comment (weshaggard) [inline eng/pipelines/apiview-review-gen-cadl.yml:48]: Not sure but doesn't cadl need to be installed first?

File: eng/scripts/Create-Apiview-Token-Cadl.ps1 [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    $tokenFilePath = Join-Path $packagePath "cadl-output"
+    $stagingReviewPath = Join-Path $OutputDir $reviewId
+    $stagingPath = Join-Path $stagingReviewPath $revisionId
+    Write-Host "Copying APIView file from '$($tokenFilePath)' to '$($stagingPath)'"
+    New-Item $stagingPath -ItemType Directory -Force
+    Copy-Item -Destination $stagingPath -Path "$tokenFilePath/*"
+}
+
+
+$reviewDetails = "$Reviews"
```
Comment (weshaggard) [inline eng/scripts/Create-Apiview-Token-Cadl.ps1:48]: Why is this extra variable needed? Isn't $Reviews already a string?

File: eng/scripts/Create-Apiview-Token-Cadl.ps1 [hunk 8]
```diff
@@ -0,0 +1,98 @@
+[CmdletBinding()]
+param (
+    [ValidateNotNullOrEmpty()]
+    [string] $Reviews,
+    [ValidateNotNullOrEmpty()]
+    [string] $WorkingDir,
+    [ValidateNotNullOrEmpty()]
+    [string] $OutputDir
+)
+
```
Comment (weshaggard) [inline eng/scripts/Create-Apiview-Token-Cadl.ps1:10]: Lets add `Set-StrictMode -Version 3`

File: eng/scripts/Create-Apiview-Token-Cadl.ps1 [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    [ValidateNotNullOrEmpty()]
+    [string] $OutputDir
+)
+
+$sparseCheckoutFile = ".git/info/sparse-checkout"
+
+
+function Sparse-Checkout($branchName, $packagePath)
+{
+    git sparse-checkout init --cone
```
Comment (weshaggard) [inline eng/scripts/Create-Apiview-Token-Cadl.ps1:16]: Doesn't this fail if you reuse on the same clone?

Comment (weshaggard) [reply eng/scripts/Create-Apiview-Token-Cadl.ps1:16]: Perhaps we need to delete the sparse checkout file before calling init again?

**Review by weshaggard — APPROVED** [review]

**Pull request merged**